src/OpenFeature/EventExecutor.cs

```diff
@@ -0,0 +1,294 @@
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using OpenFeature.Constant;
+using OpenFeature.Model;
+
+namespace OpenFeature
+{
+    internal class EventExecutor
+    {
+        private readonly Mutex _mutex = new Mutex();
+        public readonly Channel<object> EventChannel = Channel.CreateBounded<object>(1);
+        private FeatureProviderReference _defaultProvider;
+        private readonly Dictionary<string, FeatureProviderReference> _namedProviderReferences = new Dictionary<string, FeatureProviderReference>();
+        private readonly List<FeatureProviderReference> _activeSubscriptions = new List<FeatureProviderReference>();
+        private readonly SemaphoreSlim _shutdownSemaphore = new SemaphoreSlim(0);
+
+        private readonly Dictionary<ProviderEventTypes, List<EventHandlerDelegate>> _apiHandlers = new Dictionary<ProviderEventTypes, List<EventHandlerDelegate>>();
+        private readonly Dictionary<string, Dictionary<ProviderEventTypes, List<EventHandlerDelegate>>> _clientHandlers = new Dictionary<string, Dictionary<ProviderEventTypes, List<EventHandlerDelegate>>>();
+
+        public EventExecutor()
+        {
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+            this.ProcessEventAsync();
+#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+        }
+
+        internal void AddApiLevelHandler(ProviderEventTypes eventType, EventHandlerDelegate handler)
+        {
+            this._mutex.WaitOne();
```

Comment:
we should use the try-finally style for mutex usage so that we lock and release mutex regardless of any exception.

Reply:
yes, I will adapt the implementation to use a simple lock